File: tusow/src/main/java/it/unibo/coordination/tusow/api/TupleSpaceApi.java
package it.unibo.coordination.tusow.api;

import io.vertx.core.Promise;
import it.unibo.coordination.linda.core.Match;
import it.unibo.coordination.linda.core.Template;
import it.unibo.coordination.linda.core.Tuple;

import java.util.Collection;

public interface TupleSpaceApi<T extends Tuple<T>, TT extends Template<T>, K, V, M extends Match<T, TT, K, V>> extends Api {

    void createNewTuples(String tupleSpaceName, boolean bulk, Collection<? extends T> tuples, Promise<Collection<? extends T>> promise);

    void observeTuples(String tupleSpaceName, boolean bulk, boolean predicative, boolean negated, TT template, Promise<Collection<? extends M>> promise);

    void consumeTuples(String tupleSpaceName, boolean bulk, boolean predicative, TT template, Promise<Collection<? extends M>> promise);

    void getAllTuples(String tupleSpaceName, Promise<Collection<? extends T>> promise);

    void countTuples(String tupleSpaceName, Promise<Integer> promise);

}

File: tusow-service/src/main/kotlin/it/unibo/coordination/tusow/api/TupleSpaceApi.kt
package it.unibo.coordination.tusow.api

import io.vertx.core.Promise
import it.unibo.coordination.linda.core.Match
import it.unibo.coordination.linda.core.Template
import it.unibo.coordination.linda.core.Tuple

interface TupleSpaceApi<T : Tuple<T>, TT : Template<T>, K, V, M : Match<T, TT, K, V>> : Api {
    fun createNewTuples(tupleSpaceName: String, bulk: Boolean, tuples: Collection<T>, promise: Promise<Collection<T>>)
    fun observeTuples(tupleSpaceName: String, bulk: Boolean, predicative: Boolean, negated: Boolean, template: TT, promise: Promise<Collection<M>>)
    fun consumeTuples(tupleSpaceName: String, bulk: Boolean, predicative: Boolean, template: TT, promise: Promise<Collection<M>>)
    fun getAllTuples(tupleSpaceName: String, promise: Promise<Collection<T>>)
    fun countTuples(tupleSpaceName: String, promise: Promise<Int>)
}